accepted/benchmark-infrastructure.md

```diff
@@ -0,0 +1,102 @@
+unreliable environments for accurate, low-noise benchmark measurements.
+
+We already have access to one AArch64 server-grade host for interactive
+development of the Cranelift ARM backend, generously [provided by
+ARM](https://github.com/WorksOnArm/cluster/issues/204). We may be able to use
+this host for running benchmarks as well, if all stakeholders agree.
+
+We will need an x86-64 host as well; the best option is likely to rent a
+low-cost dedicated host. Details of hosting choice and funding are TBD.
+
+Relevant Bytecode Alliance-affiliated engineers would then be granted access to
+these machines and maintain a benchmark-runner infrastructure as described
+below.
+
+## Step 2: Build and host a metrics-visualization web UI
+
+We will need a frontend to present the benchmarking data that this
+infrastructure collects. Johnnie Birch and Andrew Brown of Intel have developed
+a prototype web-based frontend that records and plots performance of benchmark
+suites over time; this is the most likely candidate. As long as traffic is low
+enough, we can plan to use the benchmark-runner host(s) as web hosts as well in
+order to serve this UI. Alternately, if hosting performance becomes an issue,
+we could design a static/pregenerated-page workflow wherein the runners upload
+data to a git repository and static data is served from GitHub Pages (or
+another static host).
+
+Initially, we can populate the database for this UI with manually-initiated
+benchmarking runs on the benchmarking machines.
+
+## Step 3: Set up GitHub-integrated benchmark runner automation
+
+Once we have the machines set up, a UI established, and the ability to run
+benchmarks and upload results, we should develop a daemon that runs
+continuously on benchmark-runner machines and monitors our GitHub repositories.
+When a request for a benchmark run is initiated, the runner daemon should start
+a run; when the run is complete, it should either post the results on the
+relevant GitHub PR conversation, or include the results in its web UI, or both,
+as appropriate.
+
+The largest design question will be how this automation ensures security:
+because the runner will download a git commit and execute arbitrary code from
+that commit, we *cannot* simply execute this on every PR or every anonymous
+request. Rather, we will likely have a bot that includes an allow-list of
+trusted project members, and wait for a special GitHub comment from such a
+person to kick off a run.
+
+If any of the above becomes impractical or difficult, an intermediate
+design-point could involve a web interface that allows approved users
+(authenticated in some way) to request a run on a given git commit hash; this
+would not require any GitHub API integration. We expect, though, given prior
+art in GitHub-integrated CI-like bots (such as the Rust project's bors bot),
+that the integration should not present unforeseen problems.
+
+# Open questions
+[open-questions]: #open-questions
```

Comment:
From my experience on maintaining arewefastyet, another pain point has been availability of the service. Machines would sometimes stop running the benchmarks for Some Reason (e.g. timeout, not enough disk space on one machine, network failure,...). Then a person needs to look into it. This can lower availability, and in particular increase latency (if only some people in a given time zone have access to some machines, then the system may be unavailable for half days). I think this would be interesting to think ahead of time, and/or to mention what are the availability expectations.

Reply:
I would agree with this that maintaining continuous benchmarking, while great, does cost resources and will have hiccups. These are fine I think but it's something we'll want to go into eyes-open.

Reply:
Yes, I think the most important thing here is that it won't necessarily gate PRs: if there's a performance-sensitive thing going in and the benchmarking infrastructure is down, maybe we wait until we have numbers, but otherwise it seems like a thing we can easily enough consider optional when unavailable.